src/query/interpreter.hpp

```diff
@@ -139,6 +139,37 @@ struct QueryExtras {
   std::optional<int64_t> tx_timeout;
 };
 
+struct CurrentDB {
+  CurrentDB() = default;  // TODO: remove, we should always have an implicit default obtainable from somewhere
+                          //       ATM: it is provided by the DatabaseAccess
+                          //       future: should be a name + ptr to dbms_handler, lazy fetch when needed
+  explicit CurrentDB(memgraph::dbms::DatabaseAccess db_acc) : db_acc_{std::move(db_acc)} {}
+
+  CurrentDB(CurrentDB const &) = delete;
+  CurrentDB &operator=(CurrentDB const &) = delete;
+
+  void SetupDatabaseTransaction(std::optional<storage::IsolationLevel> override_isolation_level, bool could_commit,
+                                bool unique = false);
+  void CleanupDBTransaction(bool abort);
+  void SetCurrentDB(memgraph::dbms::DatabaseAccess new_db, bool in_explicit_db) {
+    // do we lock here?
+    db_acc_ = std::move(new_db);
+    in_explicit_db_ = in_explicit_db;
+  }
+
+  // TODO: don't provide explicitly via constructor, instead have a lazy way of getting the current/default
+  // DatabaseAccess
+  //       hence, explict bolt "use DB" in metadata wouldn't necessarily get access unless query required it.
+  std::optional<memgraph::dbms::DatabaseAccess> db_acc_;  // Current db (TODO: expand to support multiple)
```

Comment:
Can you explain a bit for what we use Gatekeeper ?

Reply:
TODO: we will need to make resources (present a talk to core devs) for all this + once close to final state (aka less cleanup) we will document within the code + add diagrams/UML